### PR TITLE
fix private log data lose problem

### PR DESCRIPTION
**Pull request by qinzuoyan**
refer to https://github.com/imzhenyu/rDSN/issues/135

### CODE REVIEW

File: src/apps/replication/lib/mutation_log.cpp [hunk 1]
```diff
@@ -615,6 +620,13 @@ decree mutation_log::max_decree(global_partition_id gpid) const
     }
 }
 
+decree mutation_log::max_commit() const
```
Comment (imzhenyu) [inline src/apps/replication/lib/mutation_log.cpp]: better to be max_on_disk_commit()

File: src/apps/replication/lib/mutation_log.cpp [hunk 2]
```diff
@@ -319,8 +315,9 @@ void mutation_log::create_new_pending_buffer()
     dassert(_pending_write == nullptr, "");
     dassert(_pending_write_callbacks == nullptr, "");
 
-    _pending_write_callbacks.reset(new std::list< ::dsn::task_ptr>);
     _pending_write = _current_log_file->prepare_log_block();
+    _pending_write_callbacks.reset(new std::list< ::dsn::task_ptr>);
+    _pending_write_max_commit = 0;
```
Comment (imzhenyu) [inline src/apps/replication/lib/mutation_log.cpp]: this should be unnecessary? why this one?

File: src/apps/replication/lib/mutation_log.cpp [hunk 3]
```diff
@@ -851,6 +869,21 @@ void mutation_log::update_max_decree_no_lock(global_partition_id gpid, decree d)
     }
 }
 
+void mutation_log::update_max_commit(decree d)
```
Comment (imzhenyu) [inline src/apps/replication/lib/mutation_log.cpp]: max commit => max on disk commit

File: src/apps/replication/lib/mutation_log.cpp [hunk 4]
```diff
@@ -1126,10 +1159,11 @@ int mutation_log::garbage_collection(replica_log_info_map& durable_decrees)
                 }
 
                 global_partition_id gpid = kv.first;
-                decree durable_decree = kv.second.max_decree;
+                decree gc_max_decree = kv.second.max_decree;
                 int64_t valid_start_offset = kv.second.valid_start_offset;
 
-                bool delete_ok_for_this_replica;
+                bool delete_ok_for_this_replica = false;
+                bool kickout_this_replica = false;
```
Comment (imzhenyu) [inline src/apps/replication/lib/mutation_log.cpp:1165]: I don't get the usage of kickout_this_replica, and its assignments are exactly the same compared to delete_ok_for_this_replica. Any particular reason to introduce this new variable?

File: src/apps/replication/lib/prepare_list.cpp [hunk 5]
```diff
@@ -138,6 +139,7 @@ error_code prepare_list::prepare(mutation_ptr& mu, partition_status status)
                 {
                     mutation_ptr mu2 = get_mutation_by_decree(d);
                     pop_min();
+                    // TODO(qinzuoyan): when will mu2 == nullptr?
```
Comment (imzhenyu) [inline src/apps/replication/lib/prepare_list.cpp]: I cannot remember a case neither - let's put an assert here.

**Comment by imzhenyu** [comment]
I think in general the changes are great:-)

File: src/apps/replication/lib/mutation_log.cpp (hunk 2 above)
Comment (qinzuoyan) [inline src/apps/replication/lib/mutation_log.cpp]: Just add the line here for code completeness, remove it here is also ok

File: src/apps/replication/lib/mutation_log.cpp (hunk 1 above)
Comment (qinzuoyan) [inline src/apps/replication/lib/mutation_log.cpp]: well, change it to max_commit_on_disk()

File: src/apps/replication/lib/mutation_log.cpp (hunk 4 above)
Comment (qinzuoyan) [inline src/apps/replication/lib/mutation_log.cpp:1165]: Yes, the refactor does not change the logic of the code.

My consider is, though kickout_this_replica here is sync with delete_ok_for_this_replica, but it is not necessary. For example, when max_decree map is empty, delete_ok_for_this_replica should be true, but kickout_this_replica may be false.

I considered the case when a log file is broken and useless, then the max_decree map is empty. But currently we assume that this case will not happen, and assert it at line 1172. So I did not handle this case in the code. But maybe someday we need to handle it....

File: src/apps/replication/lib/prepare_list.cpp (hunk 5 above)
Comment (qinzuoyan) [inline src/apps/replication/lib/prepare_list.cpp]: ok, I mean it

File: src/apps/replication/lib/mutation_log.cpp (hunk 4 above)
Comment (imzhenyu) [inline src/apps/replication/lib/mutation_log.cpp:1165]: I c.

**Pull request merged**